code/io/timer.cpp

```diff
@@ -289,17 +305,111 @@ int ui_timestamp_since(UI_TIMESTAMP stamp)
 	return timer_get_milliseconds() - stamp.value();
 }
 
-int timestamp_has_time_elapsed(int stamp, int time) {
+int timestamp_compare(TIMESTAMP t1, TIMESTAMP t2)
+{
+	Assertion(t1.isValid(), "timestamp_compare called with an invalid 't1' timestamp!");
+	Assertion(t2.isValid(), "timestamp_compare called with an invalid 't2' timestamp!");
+	if (!t1.isValid() || !t2.isValid())
+		return 0;
+
+	if (t1.isImmediate())
+	{
+		if (t2.isImmediate())
+			return 0;
+		else
+			return -1;
+	}
+	else if (t2.isImmediate())
+		return 1;
+
+	if (t1.isNever())
+	{
+		if (t2.isNever())
+			return 0;
+		else
+			return 1;
+	}
+	else if (t2.isNever())
+		return -1;
+
+	if (t1.value() == t2.value())
+		return 0;
+	else if (t1.value() > t2.value())
+		return 1;
+	else
+		return -1;
+}
+
+int ui_timestamp_compare(UI_TIMESTAMP t1, UI_TIMESTAMP t2)
+{
+	Assertion(t1.isValid(), "ui_timestamp_compare called with an invalid 't1' timestamp!");
+	Assertion(t2.isValid(), "ui_timestamp_compare called with an invalid 't2' timestamp!");
+	if (!t1.isValid() || !t2.isValid())
+		return 0;
+
+	if (t1.isImmediate())
+	{
+		if (t2.isImmediate())
+			return 0;
+		else
+			return -1;
+	}
+	else if (t2.isImmediate())
+		return 1;
+
+	if (t1.isNever())
+	{
+		if (t2.isNever())
+			return 0;
+		else
+			return 1;
+	}
+	else if (t2.isNever())
+		return -1;
+
+	if (t1.value() == t2.value())
+		return 0;
+	else if (t1.value() > t2.value())
+		return 1;
+	else
+		return -1;
+}
+
+bool timestamp_in_between(TIMESTAMP stamp, TIMESTAMP before, TIMESTAMP after)
+{
+	Assertion(stamp.isValid() && !stamp.isNever(), "timestamp_in_between was called with an Invalid or Never 'stamp' timestamp!");
```

Comment:
If there's two conditions on an assert or assertion, we really should output which statement tripped it up.

Reply:
Which statement?  Do you mean whether it was 'Invalid' or 'Never'?

Is that really necessary?  That would double the number of Assertions here.  There are already three for the three separate parameters.

Reply:
What's wrong with `Assertion(stamp.isValid() && !stamp.isNever(), "timestamp_in_between was called with an %s 'stamp' timestamp!", (stamp.isNever()) ? "isNever" : "invalid");`?

Reply:
Seems a bit convoluted, but I can add it.

Reply:
Thanks.  Imho, we make the code clear when we can, but clarity of Assertions and error messages take priority over that.